### PR TITLE
Reapply "[ubsan] Add -fsanitize-merge (and -fno-sanitize-merge) (#120…464)"

### PR DESCRIPTION
**Pull request by thurstond**
This reverts commit 2691b964150c77a9e6967423383ad14a7693095e. This reapply fixes the buildbot breakage of the original patch, by updating clang/test/CodeGen/ubsan-trap-debugloc.c to specify -fsanitize-merge (the default, which is merge, is applied by the driver but not clang_cc1).

This reapply also expands clang/test/CodeGen/ubsan-trap-merge.c.

----

Original commit message:
'-mllvm -ubsan-unique-traps' (https://github.com/llvm/llvm-project/pull/65972) applies to all UBSan checks. This patch introduces -fsanitize-merge (defaults to on, maintaining the status quo behavior) and -fno-sanitize-merge (equivalent to '-mllvm -ubsan-unique-traps'), with the option to selectively applying non-merged handlers to a subset of UBSan checks (e.g., -fno-sanitize-merge=bool,enum).

N.B. we do not use "trap" in the argument name since https://github.com/llvm/llvm-project/pull/119302 has generalized -ubsan-unique-traps to work for non-trap modes (min-rt and regular rt).

This patch does not remove the -ubsan-unique-traps flag; that will override -f(no-)sanitize-merge.

### CODE REVIEW

File: clang/test/CodeGen/ubsan-trap-merge.c [hunk 1]
```diff
@@ -1,10 +1,26 @@
...
-// shows whether -ubsan-unique-traps attaches 'nomerge' to each ubsan call.
+// The most important assertions are the attributes at the end of the file, which
+// show whether -ubsan-unique-traps and -fno-sanitize-merge attach 'nomerge'
+// to each ubsan call.
 //
-// RUN: %clang_cc1 -triple x86_64-linux-gnu -emit-llvm -fsanitize=signed-integer-overflow -O3 -mllvm -ubsan-unique-traps %s -o - -fsanitize-trap=signed-integer-overflow | FileCheck %s --check-prefix=TRAP
-// RUN: %clang_cc1 -triple x86_64-linux-gnu -emit-llvm -fsanitize=signed-integer-overflow -O3 -mllvm -ubsan-unique-traps %s -o -                                         | FileCheck %s --check-prefix=HANDLER
-// RUN: %clang_cc1 -triple x86_64-linux-gnu -emit-llvm -fsanitize=signed-integer-overflow -O3 -mllvm -ubsan-unique-traps %s -o - -fsanitize-minimal-runtime              | FileCheck %s --check-prefix=MINRT
+// N.B. although the clang driver defaults to -fsanitize-merge=undefined,
+// clang_cc1 defaults to non-merge. (This is similar to -fsanitize-recover, for
```
Comment (vitalybuka) [inline clang/test/CodeGen/ubsan-trap-merge.c:7]: how about removing `-ubsan-unique-traps` from tests?
Later we can remove it completely.

Comment (vitalybuka) [reply clang/test/CodeGen/ubsan-trap-merge.c:7]: maybe removing from tests later is also OK

**Review by vitalybuka — APPROVED** [review]

File: clang/test/CodeGen/ubsan-trap-merge.c (hunk 1 above)
Comment (thurstond) [reply clang/test/CodeGen/ubsan-trap-merge.c:7]: I am planning to write a patch that will remove -ubsan-unique-traps and update the tests accordingly.

Comment (thurstond) [reply clang/test/CodeGen/ubsan-trap-merge.c:7]: > I am planning to write a patch that will remove -ubsan-unique-traps and update the tests accordingly.

https://github.com/llvm/llvm-project/pull/120613

**Pull request merged**